Finish the CE part of the logarithmic offset RFC

mkostoevr opened the pull request:
This patchset completes the CE part of the logarithmic offset RFC:
1. accelerate the `select` with `offset` in memtx;
2. introduce the `offset` parameter of `index:pairs` method;
3. introduce the `index:offset_of` method.

Closes #8204

mkostoevr commented on src/box/memtx_tx.c:
```diff
@@ -3341,36 +3458,55 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
+	item->until = until;
+	item->until_hint = until_hint;
+
+	/*
+	 * The @a until tuple may be invisible to the counting transaction and
+	 * may be deleted later on, let's ref it to prevent its deletion until
+	 * this gap item is deleted.
+	 */
+	if (item->until)
+		tuple_ref(item->until);
```
This must be revised after the TX management is documented cause the tuple won't be deleted until it's story is, but the lifetime of stories is not clear enough to me yet.

mkostoevr replied:
The tuple is dirty, so there's a story hilding it from the deletion. And the story won't be deleted until the last gap item that uess it is deleted, but since the count tracking is performed after the ruple had been clarified, the tuple stories has `GAP_INPLACE` items in their `read_gaps`, which will only be deleted after all transaction's gap items are removed (including this count).

So it's OK to not reference the tuple here.

CuriousGeorgiy commented on src/box/memtx_tree.cc:
```diff
@@ -932,25 +975,59 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
+
+		/* Skip more visible tuples. */
+		memtx_tree_iterator_t<USE_HINT> *iterator = &it->tree_iterator;
+		while (need_skip_more_visible && res != NULL) {
```
Nit: `need_skip_more_visible != 0`.

mkostoevr replied:
Done.

CuriousGeorgiy commented on src/box/memtx_tx.c:
```diff
@@ -3309,6 +3403,30 @@ memtx_tx_inplace_gap_item_new(struct txn *txn)
+ * length in the @a key_len output parameter.
+ */
+static char *
+memtx_tx_save_key(struct txn *txn, const char *key, uint32_t part_count,
```
Nit: looks like this can be introduced as a refactoring of gap items in a separate patch.

mkostoevr replied:
Done.

CuriousGeorgiy commented on src/box/memtx_tree.cc:
```diff
@@ -965,23 +1042,60 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 	}
 
 /********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
+	if (it->skip != 0) {
```
This is out of scope of this PR, but:

Doesn't this kill the optimization? I understand there is no another way with our current isolation level, but I would argue this at least one example where one could benefit from weaker isolation levels.

mkostoevr replied:
It does in case of a big number of concurrent statements. We've discussed offline with @alyapunov and @drewdzzz about this issue and it's been desided to prefer consistency over performance with minimal amount of changes until we have real MVCC users reporting about performance problems.

But I aggree with your point in general, MVCC-oblivious count would hurt much less in some cases.

CuriousGeorgiy replied:
> But I aggree with your point in general, MVCC-oblivious count would hurt much less in some cases.

It’s not just about count, it’s also about select with offset.

Could you please open a ticket about allowing to lower the isolation level?

mkostoevr replied:
Not sure I understood the your main concern. There're two aspects of this:
1. The line you refer to creates a global `GAP_COUNT` item, which slows down any concurrent inserts and deletes until the counting transaction is finished. Do you mean to introduce an ability to prevent count gap tracking? This way a transaction that once counted items in an index may have an inconsistent value received next time.
2. It's also possible (in theory), count raw tuples without traversing existing stories (which may be slow if we have a big amount of ones) to substract invisible ones. This way the transaction will potentially see uncommitted concurrent writes and subjects for garbage collection.

mkostoevr replied:
Not sure I understood the idea. There're two aspects of this:
1. The line you refer to creates a global `GAP_COUNT` item, which slows down any concurrent inserts and deletes until the counted (selected with offset) transaction is finished. Do you mean to introduce an ability to prevent count gap tracking? This way a transaction that once counted items in an index may have an inconsistent value received next time. It's worse for select, where it can return different sets of values on each indentical request, for example:

   ```lua
   -- Index: [1, 3, 5, 7, 9]
   --[[TX1]]: select(nil, {offset = 3}) -- [7, 9]
   --[[TX2]]: insert(2)                 -- Now index: [1, 2, 3, 5, 7, 9].
   -- If TX1 doesn't create count gap item, TX2 doesn't conflict with it, so no read view for TX1.
   --[[TX1]]: select(nil, {offset = 3}) -- [5, 7, 9]
   ```

2. It's also possible (in theory), count raw tuples without traversing existing stories to substract invisible ones. This way the transaction will potentially count uncommitted concurrent writes and subjects for garbage collection. It may be OK for `index:len()`, but not for select with offset where we must only count the offset for visible tuples. The result for select is the same as above in this case.

CuriousGeorgiy replied:
I was rather referring to 2, because I don’t like the additional traversing. But giving it a thought, I guess it doesn’t worsen the situation drastically, compared to the existing MVCC overheads.

CuriousGeorgiy commented on src/box/memtx_tree.cc:
```diff
@@ -932,25 +975,59 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
-	if (it->after_data.key != NULL && res != NULL &&
-	    (type == ITER_EQ || type == ITER_REQ)) {
+	if (((it->after_data.key != NULL && is_eq) || it->skip != 0) &&
+	    res != NULL) {
```
Nit: you do some microrefactoring across this patch, like replacing `(type == ITER_EQ || type == ITER_REQ)` with `is_eq `. It would be a little simpler to review if you would try to factor them out into separate patches.

mkostoevr replied:
Done.

CuriousGeorgiy commented on src/box/memtx_tree.cc:
```diff
@@ -965,23 +1042,60 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
-	if (!key_is_full ||
-	    ((type == ITER_GE || type == ITER_LE) && !equals) ||
-	    (type == ITER_GT || type == ITER_LT))
+	if (it->skip == 0 &&
```
Nit: maybe add a `goto` instead of `it->skip == 0`?

mkostoevr replied:
Done.

CuriousGeorgiy commented on src/box/memtx_tree.cc:
```diff
@@ -932,25 +975,59 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
+	if (it->skip != 0 && res != NULL) {
+		size_t need_skip_more_visible =
+			memtx_tx_index_invisible_count_matching_until(
+				in_txn(), space, index_base, type,
```
Nit: `in_txn()` -> `txn`.

mkostoevr replied:
Done.

CuriousGeorgiy reviewed (approved):
Only reviewed `memtx: accelerate select with offset in tree` — LGTM.

lenkis commented on changelogs/unreleased/gh-8204-pairs-offset.md:
```diff
@@ -0,0 +1,4 @@
+## feature/lua
+
+* Introduced a new `offset` parameter in the `index:pairs` method. It allows to
+  skip first tuples of the iterator (gh-8204).
```
```suggestion
  skip the first tuples of the iterator (gh-8204).
```

mkostoevr replied:
Fixed, thanks!

lenkis commented on src/box/lua/schema.lua:
```diff
@@ -2509,6 +2518,23 @@ base_index_mt.count_luac = function(index, key, opts)
     return internal.count(index.space_id, index.id, itype, key);
 end
 
+-- 1-based iterator-relative index of the first matching tuple. If the such a
```
```suggestion
-- 1-based iterator-relative index of the first matching tuple. If such a
```

mkostoevr replied:
Fixed, thanks!

lenkis commented on src/box/lua/schema.lua:
```diff
@@ -2509,6 +2518,23 @@ base_index_mt.count_luac = function(index, key, opts)
 end
 
+-- 1-based iterator-relative index of the first matching tuple. If the such a
+-- tuple does not exist, returns the index it would be located at if existed.
```
```suggestion
-- tuple does not exist, returns the index that would be located at if existed.
```

mkostoevr replied:
Reworded with "the index at which it would be located" to make it clear. Please let me know if you're OK with this.

lenkis commented on src/box/memtx_tx.h:
```diff
@@ -318,6 +318,57 @@ memtx_tx_track_gap(struct txn *txn, struct space *space, struct index *index,
+}
+
+/**
+ * Record in TX manager that a transaction @a txn have counted @a index from @a
```
```suggestion
 * Record in TX manager that a transaction @a txn has counted @a index from @a
```

lenkis commented on changelogs/unreleased/gh-8204-select-offset-count-perf.md:
```diff
@@ -0,0 +1,5 @@
+## bugfix/box
+
+* Significantly improved the performance of `select` requests with the
+  `offset` specified and `count` requests. This prevents DoS conditions
```
```suggestion
  `offset` specified and with `count` requests. This prevents DoS conditions
```

mkostoevr replied:
The "`select` requests with the `offset` specified" and "`count` requests" are separated entities, performance is improved for each of them. It does not seem "with" can be applied here.

Please feel free to correct me.

mkostoevr commented:
Rebased to the current master after since the base PR (#10139) had been merged.

locker commented on src/box/memtx_tx.c:
```diff
@@ -3027,12 +3075,37 @@ memtx_tx_tuple_clarify_slow(struct txn *txn, struct space *space,
 	uint32_t res = 0;
 	struct memtx_story *story;
 	memtx_tx_foreach_in_index_tuple_story(space, index, story, {
+		/* Check the border using the cmp_def. */
+		if (until != NULL &&
+		    !memtx_tx_tuple_is_before(cmp_def, story->tuple,
+					      until, until_hint, type))
+			continue;
+
+		/* All tuples in the story chain share the same key. */
+		if (!memtx_tx_tuple_matches(key_def, story->tuple,
+					    type, key, part_count))
+			continue;
+
```
Nit: It'd be nice if this was done in one helper function because this is the second place where you call them one after another.

mkostoevr replied:
Done.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -314,6 +314,8 @@ struct tree_iterator {
 	enum iterator_type type;
 	struct memtx_tree_key_data<USE_HINT> after_data;
 	struct memtx_tree_key_data<USE_HINT> key_data;
+	/** The amount of tuples to skip after the iterator start. */
+	uint32_t skip;
```
Nit: Better name it `offset` - to match the iterator parameter.

mkostoevr replied:
Done.

locker commented on src/box/memtx_tree.cc:
```diff
@@ -886,10 +913,26 @@ memtx_tree_lookup(memtx_tree_t<USE_HINT> *tree,
+			memtx_tree_iterator_prev(tree, iterator);
 		--*offset; /* Unsigned underflow possible. */
 	}
+
+	/*
+	 * Skip the amount of tuples required. Here raw tuples are skipped,
+	 * but in case of multiple concurrent transactions, some of these
+	 * tuples might be invisible to the current one, this is handled on
+	 * the caller site.
+	 */
+	if (skip != 0) {
+		memtx_tree_skip(tree, iterator, offset, skip,
+				iterator_type_is_reverse(*type));
+	}
```
IMO better do it outside `memtx_tree_lookup` - in `tree_iterator_start`. Also, I'd inline `memtx_tree_skip` since it's small, used in one place, and having it closer to the rest of the "skip" logic would make the code easier to follow.

locker replied:
This function can't be used in the read view iterator as is because we use different types for read views. I guess we'll have to duplicate the code anyway.

mkostoevr replied:
This is a matter of overloading a bunch of functions: see the [corresponding commit in the read-view PR](https://github.com/tarantool/tarantool/pull/10478/commits/a83f5e3c42aaf27ebfdd8b3850b9df475b48eb34).

This may make code bigger, but it creates a single point of responsibility for the lookup behavior.

The same trick is done for the read-view [count implementation](https://github.com/tarantool/tarantool/pull/10357/commits/9316f0643992f428d47c216b212f83550240f7f3), so it's generic too.

locker replied:
I don't think it's a good idea because tree and its view are handled differently. For example, we don't need to do any kind of tracking in a view in contrast to a tree. If the tx manager was implemented at a different level (not in the index), it could make sense but otherwise it will probably make the code convoluted IMO. Besides, the view is implemented in EE and it should stay there for now.

mkostoevr replied:
Inlined the iterator skip into the `tree_iterator_start`.

locker commented on extra/exports:
```diff
@@ -65,6 +65,7 @@ box_index_get
 box_index_id_by_name
 box_index_iterator
 box_index_iterator_after
```
Looks like it's not used anymore. Please drop it.

mkostoevr replied:
Done.

locker commented on src/box/lua/schema.lua:
```diff
@@ -2046,8 +2049,14 @@ local function check_pairs_opts(opts, key_is_nil, level)
                 box.error(box.error.ITERATOR_POSITION, level and level + 1)
             end
         end
+        if opts.offset ~= nil then
+            offset = opts.offset
+            if type(offset) ~= "number" then
+                box.error(box.error.INVALID_OFFSET, level and level + 1)
+            end
+        end
```
We don't have a separate error code for this in `select`. I don't see a point in adding one for `pairs`. `INVALID_ARG` should do.

mkostoevr replied:
Done, the error is catched on the Lua side now (as it is for the regular select).

locker commented on src/box/lua/schema.lua:
```diff
@@ -2518,6 +2518,23 @@ base_index_mt.count_luac = function(index, key, opts)
+-- For an existing tuple, if index:index_of(key, {iterator = it}) == N, then
+-- index:pairs(nil, {iterator = it, offset = N - 1}) will start the iterator
+-- from the same tuple as index:pairs(key, {iterator = it})
+base_index_mt.index_of = function(index, key, opts)
```
Maybe name it `offset_of` - for consistency with the `offset` parameter? Also, may be make it 0-based so that it can be passed to `pairs` with a decrement?

mkostoevr replied:
This was the first thig I suggested in the RFC, but it's [been decided](https://www.notion.so/select-offset-N-count-key-log-n-6fead8b03db6409e9d09f56dd58f1006?d=028a099c41a64d249da33a06c526bfef&pvs=4#aa664802eeca45b39f24878ee9bd01a2) to stuck with this solution.

locker replied:
I read the comments in the doc. AFAIU the argument for `index_of` was to avoid confusion with `offset`, which is 0-based, while `index_of` was suggested to be 1-based because it's common in the Lua world. But `offset` is already 0-based so I don't understand why we should bother making `index_of` 1-based, thus introducing a different term.

mkostoevr replied:
Here're the discussion stages:
1. Initially the function was `offset_of` and returned a 0-based value.
2. Then it's been said we better to return 1-based values in the Lua world.
3. Then it's been said it's better to be called `index_of` instead of `offset_of` if it's 1-based.

I aggree with your point but I'm afraid of doing anything against the [RFC](https://www.notion.so/tarantool/select-offset-N-count-key-log-n-6fead8b03db6409e9d09f56dd58f1006), feel free to add a comment there.

Important thing to be noted: the product document required the [`indexof` method returning a 1-based value](https://www.notion.so/tarantool/offset-n-count-key-log-n-4aba908b48674123892474957420c85f?pvs=4#1981fb9b6757424cb961e0c1e574b1bc) (the link to the exact block may not work as expected, see the very last block of the document).

locker replied:
I talked to @unera. He says that he suggested the name `index_of` in the original design document because it's named so in Redis. He doesn't mind naming it `offset_of`. He says that making it 0-based would be better because it would be consistent with the `offset` parameter of `select`. Making it 1-based would require the user to subtract 1 from it before using it with `select`.

Another reason I don't like the name `index_of` because `box.index.foo.index_of(...)` looks weird ("index" here is used twice but has completely different meanings).

mkostoevr replied:
Renamed to `offset_of` and made it return 0-based values. Also added a no-index case to the test suite.

drewdzzz commented on src/box/memtx_tx.c:
```diff
@@ -3369,16 +3473,31 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
+	item->part_count = part_count;
+	item->key = memtx_tx_save_key(txn, key, part_count, item->short_key,
+				      sizeof(item->short_key), &item->key_len);
+	item->until = until;
```
The tuple should be referenced here.

mkostoevr replied:
Given the following statements:
1. The `until` tuple is [clarified](https://github.com/tarantool/tarantool/blob/3d984f6e0421c2442b396ddbd35ff7d4b1c1c8e6/src/box/memtx_tree.cc#L1022), so it's dirty and there's a story [hilding](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L967) it from the deletion.
   1. The tuple story has `GAP_INPLACE` item in its read_gaps.
2. And the story [won't](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L1627-L1632) be garbage collected until the last gap item that ues it is deleted.
4. The gap item will only be deleted after all gap items of the transaction are removed (including this count).

So if the count gap item is alive (and we can access its `until` member):
1. The transaction that created it is still alive.
2. The`until` tuple story is not garbage collected since it has the `GAP_INPLACE` of the transaction in it.
3. The tuple is not deleted since it has the story referencing it.

So it's OK to not reference the tuple here.

mkostoevr replied:
Given the following statements:
1. The `until` tuple is [clarified](https://github.com/tarantool/tarantool/blob/3d984f6e0421c2442b396ddbd35ff7d4b1c1c8e6/src/box/memtx_tree.cc#L1022), so it's dirty and there's a stor
y [hilding](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L967) it from the deletion.
   1. The tuple story has `GAP_INPLACE` item in its read_gaps.
2. And the story [won't](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L1627-L1632) be garbage collected until the last gap item that ues it is deleted.
4. The gap item will only be deleted after all gap items of the transaction are [removed](https://github.com/tarantool/tarantool/blob/218fc46f5f44e6e27be73c6dceaa58f1eb377f46/src/box/memtx_tx.c#L3505-L3511) (including this count).

So if the count gap item is alive (and we can access its `until` member):
1. The transaction that created it is still alive.
2. The`until` tuple story is not garbage collected since it has the `GAP_INPLACE` of the transaction in it.
3. The tuple is not deleted since it has the story referencing it.

So it's OK to not reference the tuple here.

drewdzzz replied:
When DDL happens, all the stories are deleted, including one referencing the tuple. So, even if the transaction is aborted by DDL, it's still actually alive (waits for rollback) so the count gap and the deleted tuple still can be accessed.

mkostoevr replied:
Haven't considered the DDL machinery. Thanks!

```diff
--- a/src/box/memtx_tx.c
+++ b/src/box/memtx_tx.c
@@ -340,6 +340,12 @@ memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
                            const char *key, uint32_t part_count,
                            struct tuple *until, hint_t until_hint);
 
+/**
+ * Destroy the count gap item.
+ */
+static void
+memtx_tx_count_gap_item_destroy(struct gap_item_base *base);
+
 /**
  * Helper structure for searching for point_hole_item in the hash table,
  * @sa point_hole_item_pool.
@@ -3165,6 +3165,7 @@ memtx_tx_delete_gap(struct gap_item_base *item)
                pool = &txm.nearby_gap_item_mempoool;
                break;
        case GAP_COUNT:
+               memtx_tx_count_gap_item_destroy(item);
                pool = &txm.count_gap_item_mempool;
                break;
        case GAP_FULL_SCAN:
@@ -3473,11 +3474,7 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
 }
 
 /**
- * Allocate and create count gap item. The @a until tuple's story must have
- * a gap item from the @a txn transaction or be tracked by it, so the story
- * is not deleted by the garbage collector and the tuple is not deleted (if
- * it's not NULL).
- *
+ * Allocate and create count gap item.
  * Note that in_read_gaps base member must be initialized later.
  */
 static struct count_gap_item *
@@ -3497,10 +3494,20 @@ memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
                                      sizeof(item->short_key), &item->key_len);
        item->until = until;
        item->until_hint = until_hint;
+       if (item->until != NULL)
+               tuple_ref(item->until);
 
        return item;
 }

+/**
+ * Destroy the count gap item.
+ */
+static void
+memtx_tx_count_gap_item_destroy(struct gap_item_base *base)
+{
+       struct count_gap_item *item = (struct count_gap_item *)base;
+       if (item->until != NULL)
+               tuple_unref(item->until);
+}
+
 /**
  * Allocate and create full scan gap item.
  * Note that in_read_gaps base member must be initialized later.
```

mkostoevr replied:
@drewdzzz rebased and dropped these changes since the #10160 (which removes all the global gaps including `GAP_COUNT` on DDL along with stories) is merged.

mkostoevr commented on src/box/memtx_tree.cc:
```diff
@@ -978,13 +979,12 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 		memtx_tx_track_gap(txn, space, index_base, successor, type,
-				   start_data.key,
-				   MIN(start_data.part_count, key_part_count));
+				   start_data.key, MIN(start_data.part_count,
+				   index_base->def->key_def->part_count));
```
This commit makes the next one less invasive (and thus, easier to read). I can squash it once the review is completed.

locker replied:
It isn't for me, actually. Without the context (which is in a different commit), I don't understand why this is done - this change by itself doesn't make the code any better or worse.

mkostoevr replied:
Squashed the commits into one ("memtx: accelerate select with offset in tree") since all the reviewers have watched the changes.

mkostoevr commented on src/box/memtx_tx.c:
```diff
@@ -3339,19 +3363,8 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
-						   MEMTX_TX_ALLOC_TRACKER);
-	}
-	memcpy((char *)item->key, key, item->key_len);
+	item->key = memtx_tx_save_key(txn, key, part_count, item->short_key,
+				      sizeof(item->short_key), &item->key_len);
```
See the [comment above](https://github.com/tarantool/tarantool/pull/10317#discussion_r1741899121).

mkostoevr commented on changelogs/unreleased/gh-8204-select-offset-count-perf.md:
```diff
@@ -0,0 +1,5 @@
+
+* Significantly improved the performance of `select` requests with the
+  `offset` specified and `count` requests. This prevents DoS conditions
+  previously possible (gh-8204).
```
Updated the changelog and added documentation to the commit regarding the [`index_object:count()`](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/count/), [`index_object:select()`](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/select/) and the [`difference between memtx and vinyl storage engines`](https://www.tarantool.io/en/doc/latest/platform/engines/memtx_vinyl_diff/).

mkostoevr commented on changelogs/unreleased/gh-8204-pairs-offset.md:
```diff
@@ -0,0 +1,4 @@
+## feature/lua
```
Fixed, thanks!

mkostoevr commented on src/box/index.h:
```diff
@@ -575,6 +575,13 @@ struct index_vtab {
+							const char *key,
+							uint32_t part_count,
+							const char *pos,
+							uint32_t offset);
```
Seems like a way to go if we don't want the slice check in the index virtual methods. The only alternative discussed in the RFC was introducing a new `uint32_t offset` parameter to the existing method.

I'd implement the way you suggested if @locker has no objections.

locker replied:
> Let's add a new argument `uint32_t *offset` to `create_iterator` and each index will decrease `offset` value if it can do it better than N advancements. For example, tree index will always return 0 here since it can do offset in `O(log n)`, and hash index won't touch this argument at all so we will advance the iterator manually in `box_select`.

This would obscure the function contract. I see nothing wrong in adding a new virtual method - it looks clearer to me. The situation with pagination is different because an index implementation either supports it or not so it either succeeds or raises an error.

drewdzzz replied:
OK, we decided to check slices in index, so there is no need to implement offset for other indexes outside `create_index`.
What's about separate methods - I still don't like it, but you've already discussed it in RFC, so I resolve the comment.

drewdzzz commented on src/box/index.cc:
```diff
@@ -981,6 +981,32 @@ generic_index_create_iterator(struct index *base, enum iterator_type type,
+
+	/* Skip the required amount of tuples. */
+	for (size_t i = 0; i < offset; i++) {
+		if (box_check_slice() != 0)
```
Discussed with Sergey Petrenko, he's agree with you, so I resolve this comment.

drewdzzz reviewed (approved):
Thanks for the patch, LGTM!

mkostoevr commented:
@lenkis, please take a look at the new version of the [changelogs/unreleased/gh-8204-select-offset-count-perf.md](https://github.com/tarantool/tarantool/pull/10317/files#diff-a6ec629ca26766374a8f6b8d77a7ea1643986628f0f45b6133123a24c3ed2e5e).

mkostoevr commented:
Rebased and dropped the GAP_COUNT destruction since the #10160 (which removes all the global gaps including GAP_COUNT on DDL along with stories) is merged.

xuniq commented on changelogs/unreleased/gh-8204-select-offset-count-perf.md:
```diff
@@ -0,0 +1,10 @@
+## feature/memtx
+
+* Improved performance of tree index methods: `select()` with the `offset`
+  specified and `count()` method. The underlying algorithm for these is
+  changed: the old one's time complexity was `O(n)` where `n` is the value of
+  `offset` or the amount of counted tuples. The new algorithm's complexity is
+  `O(log(size))` where `size` is the number of tuples stored in the index. So
+  now it does not depend on the value of `offset` or the amount of tuples to
+  count and it's safe to use these functions with arbitrary big offset values
+  and tuple count (gh-8204).
```
```suggestion
* Improved performance of tree index methods: `select()` with the `offset`
  specified and `count()` method. The underlying algorithm for these methods is
  changed: the old one's time complexity was `O(n)`, where `n` is the value of
  `offset` or the amount of counted tuples. The new algorithm's complexity is
  `O(log(size))`, where `size` is the number of tuples stored in the index. Now it does not depend on the `offset` value or the amount of tuples to
  count. It is safe to use these functions with arbitrary big offset values
  and tuple count (gh-8204).
```
